File: src/chain_of_responsibility/javascript/discounts.js
exports.FiveItemsDiscount = function (nextDiscount) {
    var _nextDiscount = nextDiscount;

    this.calculate = function (check) {
        if (check.getItems().length > 5) {
            return check.getTotal() * 0.1;
        }
        else {
            return _nextDiscount.calculate(check);
        }
    }

}

exports.MoreThan500Discount = function (nextDiscount) {
    var _nextDiscount = nextDiscount;

    this.calculate = function (check) {
        if (check.getTotal() > 500)
            return check.getTotal() * 0.07;
        else {
           return _nextDiscount.calculate(check);
        }
    }

}


exports.NoDiscount = function () {
    this.calculate = function (check) {
        return 0;
    }
}

File: src/chain_of_responsibility/c#/discounts.cs
namespace ChainOfResp
{
    abstract class DiscountAB
    {
        protected DiscountAB nextDiscount;

        public abstract double calculate(Check check);
    }

    class FiveOrMoreDiscount : DiscountAB
    {
        public FiveOrMoreDiscount(DiscountAB next)
        {
            nextDiscount = next;
        }

        public override double calculate(Check check)
        {
            if (check.items.Count > 5)
                return check.getTotal() * 0.1;
            else
                return nextDiscount.calculate(check);
        }

    }
    class MoreThan500Discount : DiscountAB
    {
        public MoreThan500Discount(DiscountAB next)
        {
            nextDiscount = next;
        }

        public override double calculate(Check check)
        {
            if (check.getTotal() > 500)
                return check.getTotal() * 0.07;
            else
                return nextDiscount.calculate(check);
        }

    }

    class NoDiscount : DiscountAB
    {
        public NoDiscount(DiscountAB next)
        {
            nextDiscount = null;
        }

        public override double calculate(Check check)
        {
            return 0;
        }

    }


}